src/EditorFeatures/Core/Implementation/CodeFixes/CodeFixService.cs

```diff
@@ -73,42 +73,49 @@ internal partial class CodeFixService : ForegroundThreadAffinitizedObject, ICode
             _fixAllProviderMap = ImmutableDictionary<object, FixAllProviderInfo>.Empty;
         }
 
-        public async Task<FirstDiagnosticResult> GetFirstDiagnosticWithFixAsync(
+        public async Task<DiagnosticSeverity?> GetWorstFixableSeverity(
             Document document, TextSpan range, CancellationToken cancellationToken)
         {
             if (document == null || !document.IsOpen())
             {
-                return default(FirstDiagnosticResult);
+                return null;
             }
 
             using (var diagnostics = SharedPools.Default<List<DiagnosticData>>().GetPooledObject())
             {
                 var fullResult = await _diagnosticService.TryAppendDiagnosticsForSpanAsync(document, range, diagnostics.Object, cancellationToken: cancellationToken).ConfigureAwait(false);
-                foreach (var diagnostic in diagnostics.Object)
-                {
-                    cancellationToken.ThrowIfCancellationRequested();
+                var groupTasks = diagnostics.Object.GroupBy(d => d.Severity)
+                                        .Select(severityGroup =>
+                                            Task.Factory.SafeStartNewFromAsync(
+                                                () => GetSeverity(document, severityGroup, range, cancellationToken), cancellationToken, TaskScheduler.Default));
 
-                    if (!range.IntersectsWith(diagnostic.TextSpan))
-                    {
-                        continue;
-                    }
+                var result = await Task.WhenAll(groupTasks).ConfigureAwait(false);
+                return result.Where(r => r.HasValue)
```

Comment:
You don't need the `HasValue` if you have the null check in `FirstOrDefault` right?